File: SRPTests/TestScripts/Python/ReadFromBuffer.py
# Test for reading from a buffer in a compute shader.
# This is a special compute-only test script. It relies on stuff
# from the test framework so will not work inside Syrup itself.
from SRPScripting import *

cs = ri.CompileShader("ComputeTest.hlsl", "ReadFromBuffer", "cs_4_0")

numElements = 16
input = range(0, numElements)

inputBuffer = ri.CreateBuffer(numElements, Format.R32_Float, input)
outputBuffer = ri.CreateBuffer(numElements, Format.R32_Float, None)
cs.FindResourceVariable("InBuffer").Set(inputBuffer)
cs.FindUavVariable("OutUAV").Set(outputBuffer.CreateUav())

expected = [2.0 * i + 12.0 for i in xrange(0, numElements)]

SetExpected(expected)
SetResultBuffer(outputBuffer)

def RenderFrame(context):
	context.Dispatch(cs, 1, 1, 1)

ri.SetFrameCallback(RenderFrame)

File: SRPTests/TestScripts/CS/ReadFromBuffer.cs
// Test for reading from a buffer in a compute shader.
// This is a special compute-only test script. It relies on stuff
// from the test framework so will not work inside Syrup itself.
using SRPScripting;
using System.Linq;

var cs = ri.CompileShader("ComputeTest.hlsl", "ReadFromBuffer", "cs_4_0");

int numElements = 16;
var input = Enumerable.Range(0, numElements).Select(x => (float)x);

var inputBuffer = ri.CreateStructuredBuffer(input);
var outputBuffer = ri.CreateUninitialisedBuffer(numElements * 4, 4);
cs.FindResourceVariable("InBuffer").Set(inputBuffer);
cs.FindUavVariable("OutUAV").Set(outputBuffer.CreateUav());

var expected = Enumerable.Range(0, numElements)
	.Select(i => 2.0f * i + 12.0f)
	.Cast<object>();

dynamic SetExpected = _testParams["SetExpected"];
dynamic SetResultBuffer = _testParams["SetResultBuffer"];

SetExpected(expected);
SetResultBuffer(outputBuffer);

ri.SetFrameCallback(context =>
{
	context.Dispatch(cs, 1, 1, 1);
});
